src/main/java/com/github/shyiko/mysql/binlog/json/JsonBinary.java

```diff
@@ -0,0 +1,1012 @@
+        int year = yearMonth / 13;
+        int month = yearMonth % 13;
+        int day = (int) (value >> 17) % (1 << 5); // 5 bits starting at 17th
+        int hour = (int) (value >> 12) % (1 << 5); // 5 bits starting at 12th
+        int min = (int) (value >> 6) % (1 << 6); // 6 bits starting at 6th
+        int sec = (int) (value % (1 << 6)); // 6 bits starting at 0th
+        int microSeconds = (int) (raw % (1 << 24));
+        formatter.valueDatetime(year, month, day, hour, min, sec, microSeconds);
+    }
+
+    /**
+     * Parse a {@code DECIMAL} value. The first two bytes are the precision and scale, followed by the binary
+     * representation of the decimal itself.
+     *
+     * @param length the length of the complete binary representation
+     * @param formatter the formatter to be notified of the parsed value; may not be null
+     * @throws IOException if there is a problem reading the JSON value
+     */
+    protected void parseDecimal(int length, JsonFormatter formatter) throws IOException {
+        // First two bytes are the precision and scale ...
+        int precision = reader.read();
+        int scale = reader.read();
+
+        // Followed by the binary representation (see `my_decimal_get_binary_size`)
+        int decimalLength = length - 2;
+        BigDecimal dec = AbstractRowsEventDataDeserializer.asBigDecimal(precision, scale, reader.read(decimalLength));
+        formatter.value(dec);
+    }
+
+    protected void parseOpaqueValue(ColumnType type, int length, JsonFormatter formatter)
+            throws IOException {
+        byte[] rawValue = new byte[length];
+        reader.read(rawValue);
+        formatter.valueOpaque(type, rawValue);
+    }
+
+    protected int readFractionalSecondsInMicroseconds() throws IOException {
+        return (int) readBigEndianLong(3);
+    }
+
+    protected long readBigEndianLong(int numBytes) throws IOException {
+        byte[] bytes = reader.read(numBytes);
+        long result = 0;
+        for (int i = 0; i != numBytes; i++) {
+            int b = bytes[i] & 0xFF;
+            result = (result << 8) | b;
+        }
+        return result;
+    }
+
+    protected int readUnsignedIndex(int maxValue, boolean isSmall, String desc) throws IOException {
+        long result = isSmall ? readUInt16() : readUInt32();
+        if (result >= maxValue) {
+            throw new IOException("The " + desc + " the JSON document is " + result +
+                    " and is too big for the binary form of the document (" + maxValue + ")");
+        }
+        if (result > Integer.MAX_VALUE) {
+            throw new IOException("The " + desc + " the JSON document is " + result + " and is too big to be used");
+        }
+        return (int) result;
+    }
+
+    protected int readInt16() throws IOException {
+        int b1 = reader.read() & 0xFF;
+        int b2 = reader.read();
+        return (short) (b2 << 8 | b1);
+    }
```

Comment:
This method and similar `read` methods in this class are also similar to those in `ByteArrayInputStream`, except that those other methods do not work in this case as verified by the integration tests in this PR. Although both sets of methods are intended to read little-endian values, the methods in this class properly treat negative values (e.g., like casting to `short` on [line 880](https://github.com/shyiko/mysql-binlog-connector-java/pull/119/files#diff-3a0ee09142d4a387cf95ef4fa8d188fcR880)). Changing `ByteArrayInputStream` methods was deemed to risky, so they are placed here instead. Also, because they are `protected`, clients could subclass this `JsonBinary` class and override any methods as deemed necessary, should there be a problem.